File: Difficulty: Medium/Dijkstra Algorithm/dijkstra-algorithm.cpp
// User Function Template
class Solution {
  public:
    vector<int> dijkstra(int V, vector<vector<int>> &edges, int src) {
        // Code here
        vector<vector<pair<int,int>>> adj(V);
        for(auto& vec: edges) {
            adj[vec[0]].push_back({vec[1],vec[2]});
        }
        
    
        
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
        
        
        vector<int> dist(V,INT_MAX);
        dist[src] = 0;
        
        pq.push({0,src});
        
        while(!pq.empty()) {
            pair<int,int> u = pq.top();
            pq.pop();
            for(auto& v: adj[u.second]) {
                if(dist[u.second]+v.second<dist[v.first]) {
                    dist[v.first] = dist[u.second]+v.second;
                    pq.push({dist[v.first],v.first});
                }
               
            }
        }
        return dist;
    }
};

File: Difficulty: Medium/Dijkstra Algorithm/dijkstra-algorithm.java
class Solution {
    public ArrayList<Integer> dijkstra(int V, int[][] edges, int src) {
        // code here
        
        int n = edges.length;
        
        ArrayList<ArrayList<int[]>> adjlist = new ArrayList<>();
        
        for(int i = 0;i<V;i++) {
            adjlist.add(new ArrayList<>());
        }
        
        for(int i = 0;i<n;i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            int w = edges[i][2];
            
            adjlist.get(u).add(new int[]{v,w});
            adjlist.get(v).add(new int[]{u,w});
        }
        
        
        int[] dist = new int[V];
        for(int i = 0;i<V;i++) dist[i] = Integer.MAX_VALUE;
        
        dist[src] = 0;
        
        PriorityQueue<int[]> que = 
        new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        
       
        
        que.add(new int[]{0,src});
        
        while(que.isEmpty()==false) {
            int[] uu = que.poll();
            
            int u = uu[1];
            int w = uu[0];
            
            if(w>dist[u]) continue;
            
            for(int[] vw: adjlist.get(u)) {
                int v = vw[0];
                int wv = vw[1];
                if(w+wv<dist[v]) {
                    dist[v] = w+wv;
                  
                    que.add(new int[]{dist[v],v});
                }
            }
        }
        
        
        ArrayList<Integer> ans = new ArrayList<>();
        
        for(int i: dist) ans.add(i);
        
        return ans;
        
    }
}